src/query/functions/aggregation/take.go

```diff
@@ -39,39 +39,50 @@ const (
 	TopKType = "topk"
 )
 
-type takeFunc func(values []float64, buckets [][]int) []float64
+type valueAndMeta struct {
+	val        float64
+	seriesMeta block.SeriesMeta
+}
+
+type takeFunc func(heap utils.FloatHeap, values []float64, buckets [][]int) []float64
+type takeInstantFunc func(heap utils.FloatHeap, values []float64, buckets [][]int, seriesMetas []block.SeriesMeta) []valueAndMeta
+type takeValuesFunc func(values []float64, buckets [][]int) []float64
 
 // NewTakeOp creates a new takeK operation
 func NewTakeOp(
 	opType string,
 	params NodeParams,
 ) (parser.Params, error) {
-	takeTop := opType == TopKType
-	if !takeTop && opType != BottomKType {
-		return baseOp{}, fmt.Errorf("operator not supported: %s", opType)
-	}
-
 	var fn takeFunc
+	var fnInstant takeInstantFunc
 	k := int(params.Parameter)
+
 	if k < 1 {
-		fn = func(values []float64, buckets [][]int) []float64 {
-			return takeNone(values, buckets)
+		fn = func(heap utils.FloatHeap, values []float64, buckets [][]int) []float64 {
+			return takeNone(values)
+		}
+		fnInstant = func(heap utils.FloatHeap, values []float64, buckets [][]int, seriesMetas []block.SeriesMeta) []valueAndMeta {
+			return takeInstantNone(values, seriesMetas)
```

Comment:
Can we just get rid of this `k < 1` special casing? It's a silly optimization which adds minimal performance for the additional complexity

Reply:
Completely agree.